File: java/src/tarehart/rlbot/time/GameTime.java
package tarehart.rlbot.time;

public class GameTime implements Comparable<GameTime> {

    private final long gameMillis;


    /**
     * @param gameSeconds reported by the RLBot framework
     */
    public static GameTime fromGameSeconds(double gameSeconds) {
        return new GameTime((long) (gameSeconds * 1000));
    }

    private GameTime(long gameMillis) {
        this.gameMillis = gameMillis;
    }

    public boolean isBefore(GameTime other) {
        return compareTo(other) < 0;
    }

    public boolean isAfter(GameTime other) {
        return compareTo(other) > 0;
    }

    public static GameTime from(GameTime time) {
        return new GameTime(time.gameMillis);
    }

    public long toMillis() {
        return gameMillis;
    }

    public GameTime plus(Duration duration) {
        return new GameTime(gameMillis + duration.toMillis());
    }

    public GameTime plusSeconds(double seconds) {
        return plus(Duration.ofSeconds(seconds));
    }

    public GameTime minus(Duration duration) {
        return new GameTime(gameMillis - duration.toMillis());
    }

    public GameTime minusSeconds(double seconds) {
        return minus(Duration.ofSeconds(seconds));
    }

    public static GameTime now() {
        return new GameTime(0);
    }


    @Override
    public String toString() {
        return String.format("%.2f", gameMillis / 1000.0);
    }

    @Override
    public int compareTo(GameTime other) {
        return Long.compare(gameMillis, other.gameMillis);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        GameTime that = (GameTime) o;

        return gameMillis == that.gameMillis;
    }

    @Override
    public int hashCode() {
        return (int) (gameMillis ^ (gameMillis >>> 32));
    }


}

File: src/main/java/tarehart/rlbot/time/GameTime.kt
package tarehart.rlbot.time

data class GameTime(private val gameMillis: Long) : Comparable<GameTime> {

    fun isBefore(other: GameTime): Boolean {
        return this < other
    }

    fun isAfter(other: GameTime): Boolean {
        return this > other
    }

    fun toMillis(): Long {
        return gameMillis
    }

    fun toSeconds(): Float {
        return gameMillis / 1000.0f
    }

    operator fun plus(duration: Duration): GameTime {
        return GameTime(gameMillis + duration.millis)
    }

    fun plusSeconds(seconds: Number): GameTime {
        return plus(Duration.ofSeconds(seconds))
    }

    operator fun minus(duration: Duration): GameTime {
        return GameTime(gameMillis - duration.millis)
    }

    operator fun minus(gameTime: GameTime): Duration {
        return Duration.ofMillis(gameMillis - gameTime.gameMillis)
    }

    fun minusSeconds(seconds: Double): GameTime {
        return minus(Duration.ofSeconds(seconds))
    }


    override fun toString(): String {
        return String.format("%.2f", gameMillis / 1000.0)
    }

    override fun compareTo(other: GameTime): Int {
        return java.lang.Long.compare(gameMillis, other.gameMillis)
    }

    companion object {

        /**
         * @param gameSeconds reported by the RLBot framework
         */
        fun fromGameSeconds(gameSeconds: Double): GameTime {
            return GameTime((gameSeconds * 1000).toLong())
        }

        fun zero(): GameTime {
            return GameTime(0)
        }
    }


}
